File: Typscript/Step-1 Basic/Maths/gcd.ts
// Time Complexity: O(min(a,b))

function gcd(a:number,b:number):number {
    let ans = 1;
    for(let i=1; i<=Math.min(a,b);i++) {
        if(a%i==0 && b%i==0) {
            ans = i
        }
    }
    return ans;
}

// Time Complexity: O(log(phie)(min(a,b)))
function gcgByEuclidean(a:number,b:number):number {
    while(a>0 && b>0) {
        if(a>b) a=a%b
        else b = b%a
    }

    if(a==0) return b 
    return a
}

console.log(gcd(12,9))
console.log(gcgByEuclidean(12,9))

File: Go/Step1/Maths/GCD.go
package main

import "math"

// Time Complexity: O(min(n1,n2))
func getGcd(n1, n2 int) int {
	min := math.Min(float64(n1), float64(n2))
	gcd := 0
	for i := 1; i <= int(min); i++ {
		if n1%i == 0 && n2%i == 0 {
			gcd = i
		}
	}
	return gcd
}

// Euclidean Algo
// Time Complexity: O(Math.log(phie)min(a,b))
func getGcdByEuclidean(n1, n2 int) int {

	for n1 > 0 && n2 > 0 {
		if n1 > n2 {
			n1 = n1 % n2

		} else {
			n2 = n2 % n1
		}
	}

	if n1 == 0 {
		return n2
	}
	return n1
}
